### ISSUE TITLE
iOS 14 network privacy flow refinement suggestion

### ISSUE THREAD

**Issue by robkerr** [espressif/esp-idf-provisioning-ios#27]
I saw that the library had been updated with iOS 13 and 14 keys for newer privacy prompts. I think there's one small detail that may have been overlooked for the softAP provisioning flow.

When the library begins connecting to the remote device over the network (i.e. making its http requests to 192.168.4.1), the standard iOS permission dialog is presented to the user: "[app name] would like to find and connect to devices on your local network".  This is expected on iOS 14

However the library code isn't waiting for the user to grant permission, and those initial http requests are failing, like so:

`NSLocalizedDescription=The Internet connection appears to be offline., NSErrorFailingURLStringKey=http://192.168.4.1/proto-ver, NSErrorFailingURLKey=http://192.168.4.1/proto-ver, _kCFStreamErrorDomainKey=1})`

I'm able to resolve the issue by adding the following change to the SoftAP flow, which prompts URLSession to wait for the LAN connection to be configured (i.e after the user authorizes it).  There may be other approaches...but this approach Apple suggested in their WWDC presentation on this topic: https://developer.apple.com/videos/play/wwdc2020/10110/

`ESPSoftAPTransport.swift line 50: 
   config.waitsForConnectivity = true
`
I could raise a PR but the change is pretty trivial so just thought I'd discuss it here for your consideration.

**Comment by vikas-chandra-mnnit** [comment]
@robkerr  Your observation is correct that initial request is failing because of permission. Thanks for giving your time and coming up with a resolution. 

`waitsForConnectivity ` is a session configuration and will be applied to every subsequent request. We have to check what will be the overall impact in provisioning of SoftAP after adding this change. We will analyse it soon and will inform you so that you can proceed with the PR. We highly appreciate your contribution.

**Comment by vikas-chandra-mnnit** [comment]
Hi @robkerr, I tried `waitsForConnectivity` session configuration with SoftAP provisioning. The drawback of using is that app will be stuck indefinitely waiting for connection if for some reason SoftAP disconnects from the phone. Since URLRequest will not give any error, we will not be able to inform user about the underlying cause. To verify this I disconnected my SoftAP device just before sending the Wi-Fi credentials from the app. That caused the app to remain in a loading state indefinitely in the final status screen. Without the above config, connection simply fails and app returns the appropriate error message.

**Comment by robkerr** [comment]
@vikas-chandra-mnnit yes that's what I see as well. It's a dilemma, because on iOS 14 what I also saw is that _without_ that setting, that initial http request _always_ fails because it tries to connect to 192.1.4.1 before the user allows LAN access and the iOS device is connected to that subnet and has an IP address (if they clear an Alert they can try again and it will work the 2nd time...but it's still not great UX to have that happen).  _with_ the setting, it's possible that the request never completes as you point out (e.g. if the user failed to plug in the device to begin with, then we can't know this until they have authorized LAN access).

I've been experimenting a little with this and the solution I'm currently using is to use the "wait" setting, and then run a timer, for example waiting 10 seconds for the user to resolve the permission issue. Then if 10 seconds have lapsed, and the iPhone _still_ doesn't have a 192.1.4.x subnet IP address, I assume either they didn't grant permission or there is some other technical issue...then call .disconnect() in your SDK (which will kill the ad-hoc WiFi connection, which gets them back on the Internet again). Then they can try again...for example after reminding them to plug the device into the wall and verifying it's in advertising mode, or maybe it took them a long time to think about and agree to the LAN access prompt.

Maybe something like that within the SDK could help...or if you have other suggestions how to deal with these scenarios in an elegant way that satisfies both problems I'm definitely interested.
